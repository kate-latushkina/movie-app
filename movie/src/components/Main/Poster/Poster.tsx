import React from "react";
import { ImovieItem } from "./Poster.types";

const Poster: React.FC = (props: ImovieItem) => {
  const { name, vote, description, urlImg } = props;
  return (
    <>
      <img src={urlImg} alt="poster" />
      <h3>{name}</h3>
      <span>{vote}</span>
      <p>{description}</p>
    </>
  );
};

export default Poster;
