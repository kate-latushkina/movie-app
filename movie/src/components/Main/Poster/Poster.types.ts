export type ImovieItem = {
  id: number;
  description: string;
  urlImg: string;
  name: string;
  vote: number;
  date: string;
};

export type Tmovies = ImovieItem[];
