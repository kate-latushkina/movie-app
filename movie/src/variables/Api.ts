// import { Request } from "./Request";

const API_KEY = "f37ea435a74a07c3e824148f66ca3fc7";
const API_URL = `https://api.themoviedb.org/3/discover/movie?sort_by=popularity.desc&api_key=${API_KEY}&page=1`;
const API_IMG = "https://image.tmdb.org/t/p/original";

export { API_KEY, API_URL, API_IMG };

// export const { addMovies, sentMovie, putMovie, deleteMovie } = {
//   addMovies: Request.getRequest(API_URL),
//   sentMovie: Request.postRequest(API_URL, ),
//   deleteMovie: Request.deleteRequest(API_URL),
//   putMovie: Request.putRequest(API_URL),
// };
