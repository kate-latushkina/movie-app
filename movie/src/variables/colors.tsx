const colors = {
  darkViolet: "#593158",
  darkGrey: "#424242",
  lightGrey: "#555555",
  pink: "#f65261",
  white: "#fff",
};

export default colors;
