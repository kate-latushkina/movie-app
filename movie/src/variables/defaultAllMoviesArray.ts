const defaultAllMoviesArray = [
  {
    id: 0,
    title: "",
    overview: "",
    poster_path: "",
    release_date: "2021-01-01",
    genre_ids: [1],
    vote_average: 5,
  },
];

export default defaultAllMoviesArray;
