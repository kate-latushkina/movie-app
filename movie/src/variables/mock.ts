export const CARD_PROP_MOC = {
  id: 1,
  overview: "This is description",
  poster_path: "/9kg73Mg8WJKlB9Y2SAJzeDKAnuB.jpg",
  title: "Poster",
  vote_average: 4,
  release_date: "2021",
};
