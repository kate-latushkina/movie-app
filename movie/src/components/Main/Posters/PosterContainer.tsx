import { Grid } from "@material-ui/core";
import React, { Dispatch, useEffect, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import styled from "styled-components";
import { AllMoviesActions } from "../../../state/actions/allMoviesAction";
import { setAllMovies } from "../../../state/reducers/actionCreators";
import { AllMoviesState } from "../../../state/reducers/allMoviesReducer";
import { selectAllMovies } from "../../../state/reducers/selectors";
import { addMovies } from "../../../variables/Api";
import Poster from "./components";

const Posters: React.FC = () => {
  const [movies, setMovies] = useState<any>([]);
  const allMovieDispatch = useDispatch<Dispatch<AllMoviesActions>>();

  useEffect(() => {
    addMovies.then(res => {
      const moviesRequest = allMovieDispatch(setAllMovies(res));
      return setMovies(moviesRequest);
    });
  }, []);
  console.log(movies);
  return (
    <>
      <PostersLength>20 movies found</PostersLength>

      {movies.payload ? (
        movies.payload.map(
          ({
            id,
            title,
            vote_average,
            poster_path,
            release_date,
            overview,
            genre_ids,
          }: AllMoviesState) => (
            <Poster
              key={id}
              release_date={release_date}
              title={title}
              vote_average={vote_average}
              poster_path={poster_path}
              id={id}
              overview={overview}
              genre_ids={genre_ids}
            />
          ),
        )
      ) : (
        <h2>No films have found here.</h2>
      )}
    </>
  );
};

export const PostersLength = styled(Grid)`
  width: 100%;
  margin: 10px 0 20px;
`;

export default Posters;
