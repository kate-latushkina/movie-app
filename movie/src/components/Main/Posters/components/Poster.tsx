import { Grid, makeStyles, createStyles } from "@material-ui/core";
import React, { Dispatch, useCallback, useEffect, useState } from "react";
import styled from "styled-components";
import { useDispatch } from "react-redux";
import { NavLink } from "react-router-dom";
import { ImovieItem } from "./Poster.types";
import { API_IMG } from "../../../../variables/Api";
import colors from "../../../../variables/colors";
import PosterMenu from "../../../modals/PosterMenu";
import addInputToState from "../../../../utils/addInputToState";
import { setMovie } from "../../../../state/reducers/actionCreators";
import { MovieActions } from "../../../../state/actions/movieAction";
import getGenreName from "../../../../utils/getGenreName";

const useStyles = makeStyles(() =>
  createStyles({
    posterItem: {
      maxWidth: "30%",
      marginBottom: "30px",
      position: "relative",
    },
    posterImg: {
      width: "100%",
      cursor: "pointer",
    },
    releaseDate: {
      border: `1px solid ${colors.darkGrey}`,
      borderRadius: "5px",
      color: colors.darkGrey,
      fontSize: "16px",
      padding: "5px 10px",
      height: "20px",
    },
    genreBlock: {
      display: "flex",
      flexWrap: "wrap",
      padding: 0,
    },
    genre: {
      border: `1px solid ${colors.darkGrey}`,
      borderRadius: "5px",
      color: colors.darkGrey,
      fontSize: "11px",
      padding: "4px 7px",
      listStyle: "none",
      margin: "5px 10px 0 0",
    },
    title: {
      paddingRight: "10px",
    },
  }),
);

const Poster: React.ComponentType<ImovieItem> = ({
  title,
  vote_average,
  poster_path,
  release_date,
  id,
  genre_ids,
  overview,
}: ImovieItem) => {
  const classes = useStyles();
  const [isPosterMenuShow, setPosterMenuShow] = useState(false);
  const movieDispatch = useDispatch<Dispatch<MovieActions>>();
  const correctGenre = getGenreName(genre_ids);

  const mouseLeave = useCallback(() => {
    setPosterMenuShow(false);
  }, [isPosterMenuShow]);

  const mouseEnter = useCallback(() => {
    setPosterMenuShow(true);
  }, [isPosterMenuShow]);

  const setMovieDetails = () => {
    const correctMovie = addInputToState({
      title,
      movieUrl: poster_path,
      release_date,
      id,
      genre: String(correctGenre),
      overview,
      vote_average,
    });
    movieDispatch(setMovie(correctMovie));
  };

  return (
    <Grid item xs={4} className={classes.posterItem}>
      <div onMouseLeave={mouseLeave} onMouseEnter={mouseEnter}>
        <NavLink to="/details">
          <img
            src={API_IMG + poster_path}
            alt={title}
            className={classes.posterImg}
            data-id={id}
            onClick={setMovieDetails}
            role="presentation"
          />
        </NavLink>
        {isPosterMenuShow ? (
          <PosterMenu
            title={title}
            poster_path={poster_path}
            release_date={release_date}
            id={id}
            genre_ids={genre_ids}
            overview={overview}
            vote_average={vote_average}
          />
        ) : null}
      </div>
      <MovieInfo>
        <span className={classes.title}>{title}</span>
        <span className={classes.releaseDate}>{release_date?.slice(0, 4)}</span>
      </MovieInfo>
      <span>{vote_average}</span>
      <ul className={classes.genreBlock}>
        {genre_ids
          ? correctGenre.map((genreName: string) => {
              return (
                <li className={classes.genre} key={genreName + String(id)}>
                  {genreName}
                </li>
              );
            })
          : null}
      </ul>
    </Grid>
  );
};

const MovieInfo = styled(Grid)`
  display: flex;
  justify-content: space-between;
`;

export default Poster;
