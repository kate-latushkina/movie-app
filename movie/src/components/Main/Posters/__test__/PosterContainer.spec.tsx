import React from "react";
import { shallow, ShallowWrapper } from "enzyme";
import mockResponseOnce from "jest-fetch-mock";
import PosterContainer from "../PosterContainer";
import { CARD_PROP_MOC } from "../../../../variables/mock";
import { API_URL } from "../../../../variables/Api";
import { Tmovies } from "../components/Poster.types";

const ApiRequest = (api: string) => {
  return fetch(api).then(res => res.json());
};

describe("PosterContainer component", () => {
  const component: ShallowWrapper = shallow(<PosterContainer />);

  it("should render PosterContainer component", () => {
    expect(component).toMatchSnapshot();
  });

  it("should call fetch", async () => {
    let data: Tmovies = [];
    mockResponseOnce(JSON.stringify(CARD_PROP_MOC));

    await ApiRequest(API_URL).then(res => {
      data = res.results;
      return data;
    });

    data.map(item => {
      const itemKeys = Object.keys(item).map((key: string) => {
        return expect(key).toBeDefined();
      });
      return itemKeys;
    });

    expect(Boolean([JSON.stringify(data)].length > 0)).toBeTruthy();
  });
});
