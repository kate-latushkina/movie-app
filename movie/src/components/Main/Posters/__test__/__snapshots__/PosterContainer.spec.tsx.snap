// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PosterContainer component should render PosterContainer component 1`] = `ShallowWrapper {}`;
